### PR TITLE
Improve parsing of switch expressions whith missing `}` or `]`

### PR DESCRIPTION
**Pull request by DoctorKrolic**
Closes: https://github.com/dotnet/roslyn/issues/67876

### CODE REVIEW

File: src/Compilers/CSharp/Portable/Parser/LanguageParser.cs
Comment (RikkiGibson) [inline src/Compilers/CSharp/Portable/Parser/LanguageParser.cs:89]: Did you consider using the same TerminatorState flag in both scenarios and using the same recovery strategy (abort on either an unexpected `:` or an unexpected `=>`)?

File: src/Compilers/CSharp/Portable/Parser/LanguageParser_Patterns.cs
Comment (RikkiGibson) [inline src/Compilers/CSharp/Portable/Parser/LanguageParser_Patterns.cs:551]: Bleh. I see. I take it that we don't have a fallback path where we simply parse arbitrary expressions here for recovery, and it would maybe be more disruptive to start doing that here.

**Review by RikkiGibson — COMMENTED** [review]
Done review pass. This approach seems good to me. Just had some extra tests and a follow-up question.

**Comment by RikkiGibson** [comment]
Will solicit compiler team for another review after I've signed off on the revisions..

**Review by CyrusNajmabadi — APPROVED** [review]

File: src/Compilers/CSharp/Test/Syntax/Parsing/SwitchExpressionParsingTests.cs
Comment (DoctorKrolic) [reply src/Compilers/CSharp/Test/Syntax/Parsing/SwitchExpressionParsingTests.cs:2239]: It appeared that in switch expressions we can actually afford to exit on `:` token and it will be treated as bad `=>`. So I modified the logic and added `*_Colon` case for every switch expression test. We now _not_ exit only on `=>` token in switch statements.

File: src/Compilers/CSharp/Test/Syntax/Parsing/SwitchExpressionParsingTests.cs
Comment (RikkiGibson) [inline src/Compilers/CSharp/Test/Syntax/Parsing/SwitchExpressionParsingTests.cs:1439]: bleh, I hate when the diff gets confused like this

File: src/Compilers/CSharp/Test/Syntax/Parsing/SwitchExpressionParsingTests.cs
Comment (RikkiGibson) [inline src/Compilers/CSharp/Test/Syntax/Parsing/SwitchExpressionParsingTests.cs:1514]: Would it be useful to have a test like `obj switch { { Prop: 1, { Prop: 2 }` here?

**Review by RikkiGibson — APPROVED** [review]

File: src/Compilers/CSharp/Test/Syntax/Parsing/SwitchExpressionParsingTests.cs
Comment (RikkiGibson) [reply src/Compilers/CSharp/Test/Syntax/Parsing/SwitchExpressionParsingTests.cs:2239]: Sounds good thanks!

**Review by RikkiGibson — APPROVED** [review]

**Comment by RikkiGibson** [comment]
@dotnet/roslyn-compiler for a second review

**Comment by DoctorKrolic** [comment]
Can I get a little bit of attention here?)

**Comment by DoctorKrolic** [comment]
Can I have a review here?

**Comment by RikkiGibson** [comment]
@jcouv could you please take a look?

**Comment by DoctorKrolic** [comment]
@RikkiGibson @jcouv A week has passed and still no activity here 😞

**Comment by DoctorKrolic** [comment]
@jcouv I guess you've returned to the office. Apology if not. I am just a bit worried, that this PR is sitting in a review limbo for too long...

**Review by jcouv — APPROVED** [review]
LGTM Thanks (iteration 5)

**Comment by RikkiGibson** [comment]
Thanks for the contribution @DoctorKrolic!

**Pull request merged**
